feather/server/src/systems/gamemode.rs

```diff
@@ -0,0 +1,226 @@
+use std::collections::HashMap;
+
+use base::Gamemode;
+use common::Game;
+use ecs::{SysResult, SystemExecutor};
+use quill_common::components::{
+    CanBuild, CanCreativeFly, CreativeFlying, CreativeFlyingSpeed, Instabreak, Invulnerable,
+    PreviousGamemode, WalkSpeed,
+};
+use quill_common::events::{
+    BuildingAbilityChangeEvent, CreativeFlyingEvent, FlyingAbilityChangeEvent, GamemodeEvent,
+    InstabreakChangeEvent, InvulnerabilityChangeEvent,
+};
+
+use crate::{ClientId, Server};
+use base::anvil::player::PlayerAbilities;
+
+pub fn register(systems: &mut SystemExecutor<Game>) {
+    systems.group::<Server>().add_system(gamemode_change);
+}
+
+fn gamemode_change(game: &mut Game, server: &mut Server) -> SysResult {
+    let mut may_fly_changes = HashMap::new();
+    let mut fly_changes = HashMap::new();
+    let mut instabreak_changes = HashMap::new();
+    let mut build_changes = HashMap::new();
+    let mut invulnerability_changes = HashMap::new();
+    for (
+        entity,
+        (
+            event,
+            &client_id,
+            &walk_speed,
+            &fly_speed,
+            mut may_fly,
+            mut is_flying,
+            mut instabreak,
+            mut may_build,
+            mut invulnerable,
+            gamemode,
+            prev_gamemode,
+        ),
+    ) in game
+        .ecs
+        .query::<(
+            &GamemodeEvent,
```

Comment:
I wanted to name it like this (and I did in the commands PR), but we have `CreativeFlyingEvent` and not `CreativeFlyingChangeEvent`, `SprintEvent` instead of `SprintToggleEvent`, `SneakEvent`, so I changed it to be in the same style. Forgot to change `InstabreakChangeEvent` though.